allow wildfly to use up to 4GB of heap and 4GB of permgen

cap10morgan opened the pull request:
This was defaulting to 512MB of max heap and 256MB of max permgen before and we were seeing services crash w/ out of memory errors.

The CoreOS nodes each have 15GB of RAM, but non-WildFly things like RabbitMQ will want to use some of that too.

This also bumps up the initial heap size from 64MB to 1GB. That _could_ be a problem in dev, I suppose. Maybe @tank157 has some insight on whether that's a good change or not.

Finding the optimal balance here will likely require some more experience actually running these things in production.

No card, I just saw services crashing and the out-of-memory errors in the wildfly logs.

cap10morgan committed:
[WM DC] allow wildfly to use up to 4GB of heap and 4GB of permgen

tank157 commented on wildfly@.service:
```diff
@@ -21,6 +21,7 @@ ExecStartPre=/usr/bin/docker pull ${DOCKER_REPO}:${VERSION}
 ExecStart=/bin/bash -c 'docker run -P --name ${CONTAINER} --restart=always \
   --env ADMIN_USERNAME=$(curl -s http://${COREOS_PRIVATE_IPV4}:8500/v1/kv/wildfly/admin-username?raw) \
   --env ADMIN_PASSWORD=$(curl -s http://${COREOS_PRIVATE_IPV4}:8500/v1/kv/wildfly/admin-password?raw) \
+  --env JAVA_OPTS="-Xms1024m -Xmx4096m -XX:MaxPermSize=4096m -Djava.net.preferIPv4Stack=true -Djboss.modules.system.pkgs=org.jboss.byteman -Djava.awt.headless=true" \
```
Putting on a very dusty JVM tuning hat here:

In theory we could start with a smaller size than 1GB and let it grow. These might be candidates for getting the sizes from consul so that in dev we can set them smaller and in prod bump them up. I'm also highly suspicious of a maximum perm gen space the size of the entire heap. That would leave no room for application memory. Perm gen is where class definitions go (and some other things like string literals, though that might have changed). Anyway, a big perm gen helps when loading lots of applications like we do, or reloading when the unloading process is imperfect (as it usually seems to be). But I've always seen the perm gen be at most 1/4 of the JVM memory space. That could be some pretty old thinking, but I can't see how having no space for heap or stack being a good thing.

cap10morgan replied:
Yeah, that makes sense. Maybe we'll leave the 64M starting size for the heap for now.

On the PermGen, I was assuming that it was separate from the heap, but you seem to be saying that it takes from the heap? In other words, we thought, "If these both fill up, then WildFly will be using Xmx+PermGen amount of RAM." But are you saying, "It will only use Xmx amount of RAM?"

tank157 replied:
I think it's just part of the heap allocation, but let me look into it.

cap10morgan committed:
[WM DC] revert to 64MB of initial heap size

We don't want to use too much right out of the gate, especially in dev
clusters. And this will grow up to -Xmx to meet the app's needs.

tank157 commented:
Quoting from some Oracle docs here: "The permanent generation is the area of the heap where class and method objects are stored." So yeah, unless they've changed things, the Perm Gen is a portion of the whole Heap, not in addition to.

cap10morgan commented:
@tank157 Great, thanks. New commit coming soon.

cap10morgan committed:
[WM DC] bump max heap to 8GB and max permgen down to 2GB

cap10morgan commented:
@tank157 That seem better?

tank157 commented:
Welcome. 10+ years of Java "experience" pays off some times. But does one really have experience with Java, or does Java experience you?

tank157 commented:
@cap10morgan It looks like a reasonable next step, sure. I will say there are also warning bells about an "ALLOCATE ALL THE MEMS" approach too. The warning I recall is that since each space gets bigger, including the eden (not EDN) space, the shorter, more frequent garbage collections that happen in eden space take longer, though are less frequent, but start to behave like the longer full garbage collections, which also in turn take longer.

So it's a trade off of more frequent, shorter pauses to less frequent, possibly considerably longer pauses. But now we're talking JVM tuning voodoo and my rudimentary knowledge of the garbage collection algorithms and characteristics is many years old.

tank157 commented:
But until we have evidence of that being an issue, LGTM

cap10morgan commented:
@tank157 Duly noted. I have a feeling that we'll all be becoming JVM tuning experts this year and next. :)

Thanks for the review.

tank157 commented:
I will gladly refresh my skills on that if I can skip IE XSS breakage nonsense from here on out.
